### PR TITLE
Feat: Add logout method to sdk

### PR DESCRIPTION
**Pull request by irby**
# Description

Addressing Issue #550 

Adds a log out method to the backend that the SDK can call. The motivation for this is that the Hanko session cookie is a Secure HTTP Only cookie, which cannot be modified by JavaScript for XSS protection reasons. 

# Implementation

Created a new endpoint `/logout` that's exposed on the public API routes. It calls the `UserHandler` new method `Logout` which will call upon the `SessionManager` `DeleteCookie` method. This new method has the same logic as what was in the logout endpoint in `main.go`from the quickstart project. 

In the SDK, I added a new method in `UserClient` called `logout`, which will call the `/logout` API endpoint. If this call returns an unsuccessful status code, it will throw an exception.

# Tests

To verify logic, I added unit tests for the `SessionManager.DeleteCookie` method and the `UserHandler.Logout` method. Added tests for the SDK `UserClient.logout` method that verifies successful API calls do not throw an error, while any bad status code throws the expected error.

# Todos

- ~~The logout button formatting is off. I do not know why the button appears different from the Continue button visible on the login screen.~~
  - I have figured out the resolution for this. I needed to update the common.css file to include the new `hanko-logout` element.
  
  image

- I could not figure out how to update the documentation for the frontend / backend. Are these manually updated or does Docusaurus compile it? I tried running `npm run build` but didn't make any updates.

- What should happen if token expires by the time the logout is called? The HTTP Code for an expired token will be 401. Should we treat this as an OK, logout successful, or should we still throw UnauthorizedError in the SDK?

### CODE REVIEW

**Comment by bjoern-m** [comment]
Hi @irby!

Thanks for your effort and for contributing to Hanko! About your questions:

> I could not figure out how to update the documentation for the frontend / backend. Are these manually updated or does Docusaurus compile it? I tried running npm run build but didn't make any updates.

You can run `npm run docs` and find the re-generated files under `/docs/static/jsdoc`. The backend specification can be updated by modifing `docs/static/spec/public.yaml`.

> What should happen if token expires by the time the logout is called? The HTTP Code for an expired token will be 401. Should we treat this as an OK, logout successful, or should we still throw UnauthorizedError in the SDK?

I would say that we should treat it as an OK, because otherwise whoever integrates the functionality has to check the error just to treat is the same way (e.g. redirect to a login page) and any other error, whether it is an internal server error, a bad request, etc. I would map to the "Technical Error" as the user-facing error.

I went through the code and it looked good so far, but I needed to tweak it a little bit to make it work with cross-domain setups, so I created a new branch with some suggestions, fixes, additional functionality and the updated docs. Unfortunately, I can't push it to your repository (permission denied). Do you have any idea how I can share the code with you? Haven't done this before in GitHub. Maybe I missed something, or simply do not have permission?

Regarding the logout button component, we voted in the team not to include it, because we have tentative plans that might interfere and also the components are currently meant to include the error handling. For granular functions like the logout we would recommend to use the SDK and directly call the logout function. Would that be a practical solution for you? And would it be okay to leave out the button-component for now?

Let me know what you think, and thanks again for contributing!

**Comment by irby** [comment]
Hi @bjoern-m I have added you as a contributor to my forked repo. You should be able to push changes to it. Let me know if you run into any issues. 

I am ok with leaving out the button component. I should be able to call the SDK from the quickstart project? 

Once you push your changes, I will:
- Regenerate the docs given the instructions you've provided.
- Remove the logout button component
- Call the SDK directly from the quickstart project
- Change the SDK behavior from throwing an Unauthorized error to an OK? Can you elaborate on "I would map to the "Technical Error" as the user-facing error."?

**Comment by bjoern-m** [comment]
Hi @irby,

it worked and I was able to push my branch and to create a [PR](https://github.com/irby/hanko/pull/1) in your repo! Thanks a lot! 

 >  I should be able to call the SDK from the quickstart project?

Yes, that should be possible.

> Once you push your changes, I will:
>
> * Regenerate the docs given the instructions you've provided.
> * Remove the logout button component
> * Call the SDK directly from the quickstart project

Sounds good! 

> Change the SDK behavior from throwing an Unauthorized error to an OK? Can you elaborate on "I would map to the "Technical Error" as the user-facing error."?

I just wanted to point out that there are many cases where the frontend receives a more specific response code but has no better option than display a message like "Something went wrong" which corresponds to the `TechnicalError`. I've covered this in the PR. The logout either works or it throws the technical error.

**Comment by irby** [comment]
Hi @bjoern-m sorry for the delay on this -- was traveling all last week. 

I'm currently having trouble calling the SDK from the quickstart project. Ideally, I'd like to call the `UserClient.logout` project from within the script section of the `secured.html` file. Do you have any advice on calling this function from this file?

**Comment by irby** [comment]
@bjoern-m 

Thank you for the help on calling the SDK from the quickstart project! I was stuck on that one for a bit. 😆 

I've caught up all the code with what's currently in master and had to fix up a couple of broken things, but everything should be good to go now!


Thanks,
Matt

File: quickstart/public/assets/css/common.css [hunk 1]
```diff
@@ -60,7 +60,7 @@ body {
     padding-bottom: 2rem;
 }
 
-hanko-auth, hanko-profile {
+hanko-auth, hanko-profile, hanko-logout {
```
Comment (bjoern-m) [inline quickstart/public/assets/css/common.css:63]: `hanko-logout` can be removed:

```suggestion
hanko-auth, hanko-profile {
```

Comment (irby) [reply quickstart/public/assets/css/common.css]: @bjoern-m done.

**Comment by bjoern-m** [comment]
Hi @irby!

Very nice - looks good to me now. Thanks again for the effort!

**Review by bjoern-m — APPROVED** [review]

**Comment by like-a-bause** [comment]
@bjoern-m I think you have to merge this.
@irby Thank you for your contribution!! 🥇

**Pull request merged**